services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -304,39 +323,67 @@ private boolean isDuplicateKeyError(final Throwable throwable) {
-                .withDeviceId(deviceDto.getDeviceId())
+                .withDeviceId(deviceId)
                 .document();
-        final Promise<MongoClientUpdateResult> updateDevicePromise = Promise.promise();
-        deviceDto.incrementVersionNo();
-        mongoClient.updateCollection(config.getCollectionName(), updateDeviceQuery,
-                new JsonObject().put("$set", JsonObject.mapFrom(deviceDto)), updateDevicePromise);
+        final Promise<JsonObject> updateDevicePromise = Promise.promise();
+
+        mongoClient.findOneAndReplace(config.getCollectionName(), updateDeviceQuery,
+                JsonObject.mapFrom(new DeviceDto(tenantId, deviceId, device, new Versioned<>(device).getVersion())),
+                updateDevicePromise);
+
         return updateDevicePromise.future()
-                .map(updateResult -> {
-                    if (updateResult.getDocMatched() == 0) {
-                        TracingHelper.logError(span, String.format("Device [%s] not found.", deviceDto.getDeviceId()));
-                        return OperationResult.empty(HttpURLConnection.HTTP_NOT_FOUND);
-                    } else {
-                        return OperationResult.ok(
-                                HttpURLConnection.HTTP_NO_CONTENT,
-                                Id.of(deviceDto.getDeviceId()),
-                                Optional.empty(),
-                                Optional.of(deviceDto.getVersion()));
-                    }
-                });
+                .compose(result -> Optional.ofNullable(result)
+                        .map(updated -> {
+                            span.log(String.format("successfully updated device [%s]", deviceId));
+                            return Future.succeededFuture(OperationResult.ok(
+                                    HttpURLConnection.HTTP_NO_CONTENT,
+                                    Id.of(deviceId),
+                                    Optional.empty(),
+                                    Optional.of(result.getString(RegistryManagementConstants.FIELD_VERSION))));
+                        })
+                        .orElseGet(() -> {
+                            // If the version is provided in the request, the reason for no update could be
+                            // that the device is not found or version mismatch.
+                            if (resourceVersion.isPresent()) {
+                                return findDevice(tenantId, deviceId)
+                                        .compose(foundDevice -> checkForVersionMismatchAndFail(deviceId,
+                                                foundDevice.getVersion(), resourceVersion.get()));
+                            } else {
+                                return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                        String.format("Device [%s] not found.", deviceId)));
+                            }
+                        }));
```

Comment:
looks like this is the same code as for the *delete* case, isn't it?

Reply:
I have extracted this code to a method and use it.